### PR TITLE
add buffer function

### PR DESCRIPTION
**Pull request by alex-s-gardner**

### CODE REVIEW

File: src/Extents.jl [hunk 1]
```diff
@@ -198,6 +198,18 @@ intersection(obj1::Nothing, obj2::Nothing) = nothing
 intersection(obj1, obj2) = intersection(extent(obj1), extent(obj2))
 intersection(obj1, obj2, obj3, objs...) = intersection(intersection(obj1, obj2), obj3, objs...)
 
+"""
+    buffer(ext::Extent; buff = (X=0, Y=0))
+
+buffer `Extent` in X and Y by values supplied in `buff` named tuple
+"""
+function buffer(ext::Extent; buff = (X=0, Y=0))
```
Comment (rafaqz) [inline src/Extents.jl:206]: Probably better to have no default.

File: src/Extents.jl [hunk 2]
```diff
@@ -198,6 +198,18 @@ intersection(obj1::Nothing, obj2::Nothing) = nothing
...
+    x_new = ext.X .+ (-buff.X, +buff.X);
+    y_new = ext.Y .+ (-buff.Y, +buff.Y);
+    Extent(X=x_new, Y=y_new)
```
Comment (rafaqz) [inline src/Extents.jl:209]: Extent can have fields wirh any name at all, so this has to be generic.

I think you need to get the keys from the extent type, then map over them checking `haskey` on the buffer and then buffer or leave the same.

File: test/runtests.jl [hunk 3]
```diff
@@ -72,3 +72,9 @@ end
     @test Extents.intersection(nothing, nothing) === nothing
     @test Extents.intersection(nothing, b) === nothing
 end
+
+@testset "buffer" begin
+    a = Extent(X=(0.1, 0.5), Y=(1.0, 2.0))
```
Comment (rafaqz) [inline test/runtests.jl:77]: We need to test this with :

- arbitrary names
- some names missing (dont get padded)
- datetime as the pad

**Review by rafaqz — COMMENTED** [review]
Thanks! 

Mostly we need to make this more generic and flexible. Even in geospatial data there is Z. But generally dimensions can have any name at all, and you may not want to buffer all of them.

File: src/Extents.jl (hunk 1 above)
Comment (rafaqz) [inline src/Extents.jl:206]:
```suggestion
function buffer(ext::Extent{K}, buff) where K
```

File: src/Extents.jl (hunk 2 above)
Comment (rafaqz) [inline src/Extents.jl:209]:
```suggestion
    Extent{K}(vals)
```

File: test/runtests.jl (hunk 3 above)
Comment (alex-s-gardner) [reply test/runtests.jl:77]: done!

Comment (rafaqz) [reply test/runtests.jl:77]: Thanks, but I don't see any DateTime in there ;)

**Comment by rafaqz** [comment]
Have your checked this with `@btime` ? It should only take a few nanoseconds. Checking with `@profview` may also show some problems.

**Comment by alex-s-gardner** [comment]
Note that buffer only works if coordinates are sorted: https://github.com/rafaqz/Extents.jl/pull/6

**Comment by rafaqz** [comment]
This is an open question because reversed can mean crossing the date line or similar.

Is it wrong in that case?

**Comment by alex-s-gardner** [comment]
> This is an open question because reversed can mean crossing the date line or similar.

I don't think any of the Extents math will work if you cross the date line.

**Comment by alex-s-gardner** [comment]
Execute times seem reasonable
```julia
julia> c = Extent(X=(0.1, 0.5), Y=(1.0, 2.0), Ti=(DateTime(2000, 1, 1), DateTime(2020, 1, 1)));

julia> @btime Extents.buffer(c, (X=2, Y=2))
  16.741 ns (1 allocation: 64 bytes)
Extent(X = (-1.9, 2.5), Y = (-1.0, 4.0), Ti = (DateTime("2000-01-01T00:00:00"), DateTime("2020-01-01T00:00:00")))

julia> @btime Extents.buffer(c, (X=2, Y=2, Ti=Year(1)))
  912.028 ns (36 allocations: 1.41 KiB)
```

**Comment by rafaqz** [comment]
Tryvto use $ on the variables... any allocation here is not reasonable

**Comment by alex-s-gardner** [comment]
> Try to use $ on the variables... any allocation here is not reasonable

no allocation for X/Y but there is for DateTime 
```julia
julia> @btime Extents.buffer($c, (X=2, Y=2))
  1.334 ns (0 allocations: 0 bytes)
Extent(X = (-1.9, 2.5), Y = (-1.0, 4.0), Ti = (DateTime("2000-01-01T00:00:00"), DateTime("2020-01-01T00:00:00")))

julia> @btime Extents.buffer($c, (X=2, Y=2, Ti=Year(1)))
  876.647 ns (36 allocations: 1.41 KiB)
Extent(X = (-1.9, 2.5), Y = (-1.0, 4.0), Ti = (DateTime("1999-01-01T00:00:00"), DateTime("2021-01-01T00:00:00")))
```

**Comment by rafaqz** [comment]
1.3ns seems right ;)

But yes bit of a problem with DateTime, probably worth a look with `@profview` and `Chulhu.@descend` to see what the type instability is from.

Possibly its the mixed types? You could try other mixed types and see what happens. We do need to be robust to that.

**Comment by alex-s-gardner** [comment]
> Possibly its' the mixed types? Try to use `map(+, ext[k], (-buff[k], +buff[k])) ` instead of the broadcast, `map` is type stable with mixed tuples.

this cuts the time in half but we still have considerable allocations 
```julia
buff = (X=2, Y=2, Ti=Year(1))
@btime Extents.buffer($c, $buff)
567.027 ns (30 allocations: 1.12 KiB)
```

**Comment by alex-s-gardner** [comment]
DateTime alone is not the issue:

```julia

julia> @btime DateTime(2000, 1, 1) - Year(1)

  0.750 ns (0 allocations: 0 bytes)
1999-01-01T00:00:00

julia> Ti = DateTime(2000, 1, 1);

julia> @btime $Ti - Year(1)
  22.967 ns (0 allocations: 0 bytes)
1999-01-01T00:00:00

julia> @btime map(+, (DateTime(2000, 1, 1),  DateTime(2020, 1, 1)),  (Year(1), - Year(1)))
  0.750 ns (0 allocations: 0 bytes)
(DateTime("2001-01-01T00:00:00"), DateTime("2019-01-01T00:00:00"))

julia> Ti = (DateTime(2000, 1, 1), DateTime(2020, 1, 1));

julia> buff = (Year(1), -Year(1));

julia> @btime map(+, $Ti, $buff)
  49.096 ns (0 allocations: 0 bytes)
(DateTime("2001-01-01T00:00:00"), DateTime("2019-01-01T00:00:00"))
```

**Comment by rafaqz** [comment]
Ok looks like `k` is not constant propagating.

Try `Base.@assume_effects :foldable buffer(...`

Otherwise we need to manually force `K` as `Val{K}()` by doing `map(map(Val, K)) do k`

then unwrap it with `ext[_unwrap(k)]`. 

where

```julia
_unwrap(::Val{K}) where K = K
```

**Comment by alex-s-gardner** [comment]
> Try `Base.@assume_effects :foldable buffer(...`

no luck with `Base.@assume_effects`

```julia
julia> c = Extent(X=(0.1, 0.5), Y=(1.0, 2.0), Ti=(DateTime(2000, 1, 1), DateTime(2020, 1, 1)));

julia> buff = (X=2, Y=2, Ti=Year(1));

julia> @btime Extents.buffer($c, $buff)
  568.842 ns (30 allocations: 1.12 KiB)
Extent(X = (-1.9, 2.5), Y = (-1.0, 4.0), Ti = (DateTime("1999-01-01T00:00:00"), DateTime("2021-01-01T00:00:00")))
```

**Comment by rafaqz** [comment]
Ok you need to force K to constprop with Val as above

Its a nice trick to know ;)

**Comment by alex-s-gardner** [comment]
> Otherwise we need to manually force `K` as `Val{K}()` by doing `map(map(Val, K)) do k`
> 
> then unwrap it with `ext[_unwrap(k)]`.
> 
> where
> 
> ```julia
> _unwrap(::Val{K}) where K = K
> ```

I'm having trouble following?  It's not clear to me what Val is in this case.

**Comment by rafaqz** [comment]
`Val` is a singleton type with one parameyer defined in base to do this stuff

**Comment by alex-s-gardner** [comment]
OK, that's like magic to me... now I need to go try to understand why that works.

```julia
julia> @btime Extents.buffer($c, $buff)
  45.542 ns (0 allocations: 0 bytes)
Extent(X = (-1.9, 2.5), Y = (-1.0, 4.0), Ti = (DateTime("1999-01-01T00:00:00"), DateTime("2021-01-01T00:00:00")))
```

**Comment by rafaqz** [comment]
Its forcing to compile a separate method for each call to the anonymous function, where the NameTuple key is known at compile time and const props so the return type us also knoen.

Otherwise `k` is just a runtime Symbol - and indexing the NamedTuple has a Union return type.

**Comment by rafaqz** [comment]
Ok seems there was already an `_unwrap` function, and the tests need the Dates module in Project.toml

File: src/Extents.jl [hunk 4]
```diff
@@ -198,10 +198,34 @@ intersection(obj1::Nothing, obj2::Nothing) = nothing
...
+"""
+    buffer(ext::Extent, buff::NamedTuple)
+
+buffer `Extent` by corresponding name-pair values supplied in `buff` NamedTuple.
+
+# Examples
+```julia-repl
+julia> ext = Extent(X = (1.0, 2.0), Y = (3.0, 4.0))
+Extent(X = (1.0, 2.0), Y = (3.0, 4.0))
+julia> ext = Extents.buffer(ext, (X=1,Y=3))
```
Comment (rafaqz) [inline src/Extents.jl:210]:
```suggestion
julia> ext_buffered = Extents.buffer(ext, (X=1, Y=3))
```

**Comment by rafaqz** [comment]
Thanks!

**Comment by alex-s-gardner** [comment]
@rafaqz tank you for the hand holding!

**Comment by rafaqz** [comment]
No worries, I do enjoy setting unreasonably high performance standards 😂

**Pull request merged**